tools/stadium.c

```diff
@@ -0,0 +1,154 @@
+#define BASESIZE 24
+// ASCII "N64PS3" header
+#define N64PS3SIZE 6
+// base + N64PS3 + CRC
+#define HEADERSIZE (BASESIZE + N64PS3SIZE + 2)
+// Checksum every half-bank
+#define CHECKSIZE 0x2000
+
+typedef enum Base { BASE_NO, BASE_US, BASE_EU, BASE_DBG } Base;
+
+uint8_t us_base[BASESIZE] = {'b', 'a', 's', 'e',
+	0x01, 0x00, 0xBF, 0x6B, 0x40, 0x11, 0x00, 0x22, 0x00, 0x3A,
+	0xF3, 0x38, 0x18, 0xFF, 0xFF, 0x0F, 0x07, 0x10, 0x68, 0x07};
+
+uint8_t eu_base[BASESIZE] = {'b', 'a', 's', 'e',
+	0x01, 0x01, 0x1e, 0xCF, 0x00, 0x10, 0x00, 0x00, 0x00, 0x0C,
+	0xA3, 0x38, 0x00, 0xFF, 0xFF, 0x07, 0x00, 0x00, 0x00, 0x14};
+
+uint8_t dbg_base[BASESIZE] = {'b', 'a', 's', 'e',
+	0x01, 0x00, 0x07, 0x82, 0x40, 0x10, 0x00, 0x22, 0x00, 0x3A,
+	0xE3, 0x38, 0x00, 0xFF, 0xFF, 0x07, 0x07, 0x10, 0x68, 0x06};
+
+uint8_t n64ps3[N64PS3SIZE] = {'N', '6', '4', 'P', 'S', '3'};
+
+static void usage(void) {
+	fprintf(stderr, "Usage: stadium [-h|--help] [-b|--base us|eu|dbg] romfile\n");
+}
+
+void parse_args(int argc, char *argv[], Base *b) {
+	struct option long_options[] = {
+		{"base", required_argument, 0, 'b'},
+		{"help", no_argument, 0, 'h'},
+		{0}
+	};
+	for (int opt = 0; opt != -1;) {
+		switch (opt = getopt_long(argc, argv, "hb:", long_options)) {
+		case 'h':
+			usage();
+			exit(0);
+			break;
+		case 'b':
+			*b = !strcmp(optarg, "us") ? BASE_US :
+				!strcmp(optarg, "eu") ? BASE_EU :
+				!strcmp(optarg, "dbg") ? BASE_DBG :
+				BASE_NO;
+			break;
+		case 0:
+		case -1:
+			break;
+		default:
+			usage();
+			exit(1);
+			break;
+		}
+	}
+}
+
+#define SET2(file, off, v) do { \
+	file[(off) + 0] = (uint8_t)(((v) & 0xFF00) >> 8); \
+	file[(off) + 1] = (uint8_t)(((v) & 0x00FF) >> 0); \
+} while (0)
+
+void calculate_checksums(uint8_t *file, int filesize, Base base) {
+	// Initialize the header with the appropriate base data
+	uint8_t header[HEADERSIZE] = {0};
+	if (base == BASE_US) {
+		memcpy(header, us_base, BASESIZE);
+	} else if (base == BASE_EU) {
+		memcpy(header, eu_base, BASESIZE);
+	} else if (base == BASE_DBG) {
+		memcpy(header, dbg_base, BASESIZE);
+	}
+	memcpy(header + BASESIZE, n64ps3, N64PS3SIZE);
+
+	int NUMCHECKS = filesize / CHECKSIZE;
+	int DATASIZE = HEADERSIZE + NUMCHECKS * 2; // 2 bytes per checksum
+	int ORIGIN = filesize - DATASIZE; // Stadium data goes at the end of the file
+
+	// Clear global checksum
+	SET2(file, GLOBALOFF, 0x0000);
+
+	// Clear Stadium data (this should be free space anyway)
+	memset(file + ORIGIN, 0x00, DATASIZE);
+
+	// Fill Stadium header values
+	memcpy(file + ORIGIN, header, HEADERSIZE);
+
+	// Calculate the half-bank checksums
+	for (int i = 0; i < NUMCHECKS; i++) {
+		uint16_t checksum = 0xFEFE;
+		for (int j = 0; j < CHECKSIZE; j++) {
+			checksum += file[i * CHECKSIZE + j];
+		}
+		SET2(file, ORIGIN + HEADERSIZE + i * 2, checksum);
+	}
+
+	// Initialize the CRC table
+	uint16_t crc_table[256] = {0};
+	for (int i = 0; i < 256; i++) {
+		uint16_t c = (uint16_t)i;
+		uint16_t t = 0x0000;
```

Comment:
Either call it `crc` or `rem` (remainder), as that's pretty much what's happening here, you're precalculating the crc for a set of values to match the size of the values you're reading later. I guess it could be incorporated into the same loop so the table doesn't have to be built, but here's hoping a smart compiler will realize it's static.

Reply:
I went with `rem` since `crc` below is the actual single CRC.